File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] cand, int tar) {
        List<List<Integer>> ans=new ArrayList<>();
        check(0,cand,tar,ans,new ArrayList<Integer>());
        return ans;
    }

    public void check(int n,int []cand,int k,List<List<Integer>> ans,ArrayList<Integer> demo){
        if(k==0){
            ans.add(new ArrayList<>(demo));
            return;
            }
        if(n==cand.length){
            return;
        }

        if(cand[n]<=k){
            demo.add(cand[n]);
            check(n,cand,k-cand[n],ans,demo);
            demo.remove(demo.size()-1);
        }
        check(n+1,cand,k,ans,demo);
    }
}

File: 0039-combination-sum/0039-combination-sum.cs
public class Solution {
    public IList<IList<int>> CombinationSum(int[] cand, int tar) {
        IList<IList<int>> ans=new List<IList<int>>();
        Check(0,cand,tar,ans,new List<int>());
        return ans;
    }

    public void Check(int n,int[] cand,int k,IList<IList<int>> ans,List<int> demo){
        if(k==0){
            ans.Add(new List<int>(demo));
            return;
            }
        if(n==cand.Length){
            return;
        }

        if(cand[n]<=k){
            demo.Add(cand[n]);
            Check(n,cand,k-cand[n],ans,demo);
            demo.RemoveAt(demo.Count-1);
        }
        Check(n+1,cand,k,ans,demo);
    }
}